### ISSUE TITLE
`lein compile` may print out info messages with *print-dup* enabled

### ISSUE THREAD

**Issue by hyPiRion** [technomancy/leiningen#2082]
If you do `lein compile` directly on a project where you don't have the dependencies downloaded, the dependencies are downloaded with `*print-dup*` enabled like so:

```
"Retrieving" "rome/rome/0.9/rome-0.9.pom" "from" "central"
```

We should have a way to enable/disable `*print-dup*` when doing `eval-in-project`. The easiest is probably to have a dynamic var which you can bind. That should also be backwards compatible.

**Comment by hyPiRion** [comment]
This issue is rather straightforward and only requires some basic knowledge of Clojure and dynamic vars.

More specifically, we need to add in a dynamic var in `eval.clj` in leiningen-core, which enables *print-dup* when set. This should be set instead of `*print-dup*` in the compile task.

Feel free to ask for help from me if you want to contribute.

**Comment by jolson88** [comment]
I think I'll go ahead and take a look this weekend if that's okay. I've been looking for some projects to contribute to (especially wanting to work more with Clojure). I'll go ahead and ask questions if I have them. I'll look at getting familiar with eval.clj in leiningen-core tomorrow. 

BTW, thanks for having the Newbie labels! It's an absolutely fantastic way to find ways to get started contributing on a project :). Wish more projects had this.

**Comment by hyPiRion** [comment]
@jolson88: Thanks! I'm going on vacation for a couple of days tomorrow, so I may be a bit slow to respond. But I believe the issue should be relatively straightforward.

**Comment by jolson88** [comment]
So yes, doing this should be straightforward, but looking into this more I'm wondering if it's the right way forward. I started this research based on trying to understand what would be a breaking change or not before implementing the fix. My concern would be that the usage of *print-dup* seems to be causing several issues and could turn into an ongoing game of whack-a-mole. 

As I'm not familiar enough with how some of these functions are used in the greater scope of lein, I'm going to post these thoughts/questions here for guidance. 

The first introduction of *print-dup* in lein was on February 4th in commit 99d53683ebb953ef131b967c0eeef5f4490dd42a. The commit states it was to resolve reflection warnings. That location of *print-dup* was removed two days later with 4383caecce94482a57a25146bdd236e89b20f75f as it looks like it was "causing some havoc". In the process, the binding of *print-dup* was moved to compile.clj. 

The move to compile.clj though caused lein-environ to break, as shown in #2096 (and other projects doing :aot compilation as seen linking to the issue). It also looks to have caused the issues outlined in this bug too perhaps. 

So given the regressions (and possible future regressions from whack-a-mole), should the code continue to go down the route of utilizing *print-dup*, or should a different way be looked at? So I could definitely use some context from folks with more lein contrib experience :).

**Comment by hyPiRion** [comment]
That's a good timeline of the history of the `*print-dup*`. If it's of any value, I decided to elaborate a bit more what the exact issues were in each part, for context.

The issue at hand here is that we would like to have some option to specify whether `*print-dup*` should be enabled for the forms passed to `eval-in-project`. The reason for this is to ensure that metadata and similar things will be serialised, causing type hints to be present in the forms passed in. This is what 99d53683ebb953ef131b967c0eeef5f4490dd42a attempted to fix, by enabling `*print-dup*` by default.

However, many plugins use the project map as part of the form passed to `eval-in-project`. Some of the project map values, like the `:repositories` key, has a metadata entry that refers to a Leiningen function. As functions cannot be read back in, even with `*print-dup*` enabled, this causes the "havoc" that was mentioned in 4383caecce94482a57a25146bdd236e89b20f75f.

To solve this, we decided to instead enable `*print-dup*` in the functions that needed this functionality. This usually works fine, but plugins may hook themselves into functions in Leiningen itself. Additionally, some tasks may need to run `:prep-tasks` before they are actually evaluated. Both of these will also be affected by setting `*print-dup*` to true before calling `eval-in-project`. It's easy to see if we run this in a repl:

```clj
(binding [*print-dup* true]
  (println "hello" "world")
  (pr-str ^:foo []))
"hello" "world"
"^#=(clojure.lang.PersistentArrayMap/create {:foo true}) []"
```

Although we get the expected result back (something that is `*print-dup*`, don't worry about how it looks like), we've also caused `println` to print out `"hello" "world"` instead of the usual `hello world`.

---

Now, how do we fix this? I believe the easiest way is to introduce a dynamic var inside "leiningen-core/eval.clj", which could be set by the caller. When `eval-in-project` is called, it will set `*print-dup*` to the dynamic var, but only when it prints the form.

As a small example on how I imagine it (extremely simplified, but I think you get the idea):

Instead of doing it like this, as we do today:

```clj
;; In leiningen.core.eval
(defn eval-in-project [project form]
  (println "eval-in-project!")
  (pr-str form))

;; In some other namespace
(defn my-task [...]
  (binding [*print-dup* true]
    (eval/eval-in-project project form)))
```

We can do it like this:

```clj
;; In leiningen.core.eval
(def ^:dynamic *eval-print-dup* false)

(defn eval-in-project [project form]
  (println "eval-in-project!")
  (binding [*print-dup* *eval-print-dup*]
    (pr-str form)))

;; in another namespace

(defn my-task [...]
  (binding [eval/*eval-print-dup* true]
    (eval/eval-in-project project form)))
```

This will ensure that `*print-dup*` is set when the form is generated, and only when forms are generated. It won't affect e.g. the `(println "eval-in-project!")` in the example.

There are some edge cases here, but I believe that very few `eval-in-project`-tasks will need to use `*print-dup*`, and so I _think_ this should be fine in practice (no big whack-a-moles). Also I'm not sure if this is easy to introduce properly without causing breakage, and this solution will be strictly better than the current one.

I _think_ that should answer all your questions, or at least point you in the right direction. Also, sorry for the wall of text here, I got a bit carried away =)

**Comment by jolson88** [comment]
Fantastic! Thanks for the more context 😄. That should give me plenty to put together a pull request. If I run into more questions, I'll let you know 👍.

Sent from my iPhone

On Mar 19, 2016, at 2:38 PM, hyPiRion <notifications@github.com<mailto:notifications@github.com>> wrote:


That's a good timeline of the history of the *print-dup*. If it's of any value, I decided to elaborate a bit more what the exact issues were in each part, for context.

The issue at hand here is that we would like to have some option to specify whether *print-dup* should be enabled for the forms passed to eval-in-project. The reason for this is to ensure that metadata and similar things will be serialised, causing type hints to be present in the forms passed in. This is what 99d5368<https://github.com/technomancy/leiningen/commit/99d53683ebb953ef131b967c0eeef5f4490dd42a> attempted to fix, by enabling *print-dup* by default.

However, many plugins use the project map as part of the form passed to eval-in-project. Some of the project map values, like the :repositories key, has a metadata entry that refers to a Leiningen function. As functions cannot be read back in, even with *print-dup* enabled, this causes the "havoc" that was mentioned in 4383cae<https://github.com/technomancy/leiningen/commit/4383caecce94482a57a25146bdd236e89b20f75f>.

To solve this, we decided to instead enable *print-dup* in the functions that needed this functionality. This usually works fine, but plugins may hook themselves into functions in Leiningen itself. Additionally, some tasks may need to run :prep-tasks before they are actually evaluated. Both of these will also be affected by setting *print-dup* to true before calling eval-in-project. It's easy to see if we run this in a repl:

(binding [*print-dup* true]
  (println "hello" "world")
  (pr-str ^:foo []))
"hello" "world"
"^#=(clojure.lang.PersistentArrayMap/create {:foo true}) []"

Although we get the expected result back (something that is *print-dup*, don't worry about how it looks like), we've also caused println to print out "hello" "world" instead of the usual hello world.

________________________________

Now, how do we fix this? I believe the easiest way is to introduce a dynamic var inside "leiningen-core/eval.clj", which could be set by the caller. When eval-in-project is called, it will set *print-dup* to the dynamic var, but only when it prints the form.

As a small example on how I imagine it (extremely simplified, but I think you get the idea):

Instead of doing it like this, as we do today:

;; In leiningen.core.eval
(defn eval-in-project [project form]
  (println "eval-in-project!")
  (pr-str form))

;; In some other namespace
(defn my-task [...]
  (binding [*print-dup* true]
    (eval/eval-in-project project form)))

We can do it like this:

;; In leiningen.core.eval
(def ^:dynamic *eval-print-dup* false)

(defn eval-in-project [project form]
  (println "eval-in-project!")
  (binding [*print-dup* *eval-print-dup*]
    (pr-str form)))

;; in another namespace

(defn my-task [...]
  (binding [eval/*eval-print-dup* true]
    (eval/eval-in-project project form)))

This will ensure that *print-dup* is set when the form is generated, and only when forms are generated. It won't affect e.g. the (println "eval-in-project!") in the example.

There are some edge cases here, but I believe that very few eval-in-project-tasks will need to use *print-dup*, and so I think this should be fine in practice (no big whack-a-moles). Also I'm not sure if this is easy to introduce properly without causing breakage, and this solution will be strictly better than the current one.

I think that should answer all your questions, or at least point you in the right direction. Also, sorry for the wall of text here, I got a bit carried away =)

—
You are receiving this because you were mentioned.
Reply to this email directly or view it on GitHub<https://github.com/technomancy/leiningen/issues/2082#issuecomment-198788088>

________________________________

This e-mail message is authorized for use by the intended recipient only and may contain information that is privileged and confidential. If you received this message in error, please call us immediately at (425) 590-5000 and ask to speak to the message sender. Please do not copy, disseminate, or retain this message unless you are the intended recipient. In addition, to ensure the security of your data, please do not send any unencrypted credit card or personally identifiable information to this email address. Thank you.

**Comment by 3duard0** [comment]
@hyPiRion I didn't know if jolson88 was still working on this issue, so I decided to give a try.
I hope everything is okay, I don't have much experience on opensource.

**Comment by jolson88** [comment]
Thanks 3duard0! unfortunately, I've had no time. So I'm glad you picked it up :).

**Comment by hyPiRion** [comment]
Fixed by #2082
